[LoopUnroll] Structural cost savings analysis for full loop unrolling

lucas-rami opened the pull request:
_This PR was originally meant to optimize a specific nested loop structure when targeting the AMDGPU backend. Following feedback from reviewers, the contribution was moved to the target-independent part of the loop-unrolling pass. In the process, the scope of the contribution increased non-trivially to work well with the current pass design._

When the loop unroll pass deems full unrolling of a loop possible, it checks whether the estimated full unroll cost is lower that a pre-determined threshold. When this check fails, the pass attempts to run a costly iteration-by-iteration analysis to obtain much more precise unrolled/rolled costs for the loop, then uses those results to boost the threshold by a certain amount. Due to the analysis's complexity it can only run on innermost loops with small trip count, preventing full unrolling of loops that seem too costly in other cases.

This introduces an additional simpler analysis that runs when the former analysis fails and which can work on outer loops independently of trip count. The analysis attempts to come up with a more accurate unrolled cost for the input loop to later compare against the threshold. It does so by identifying instructions that will be removable in the unrolled form of the loop (so-called "known instructions") and deducting their cost from the previously established full unroll cost. Additionally, the analysis derives cost savings from the existence of subloops whose trip count would become runtime-independent following full unrolling of an outer loop. Similarly, more cost savings can come from the target of conditional branches becoming known at runtime. If the overall unrolled cost of the loop was lowered enough by the analysis, the loop ends up fully unrolled.

## Original PR description
<details>

AMDGPU TTI's unrolling preferences disallow unrolling loops with a runtime trip count. However, it is common enough for an inner loop's bounds to only depend on an outer loop's induction variable as well as constants, for example when iterating over half of a matrix.

```cpp
for (size_t i = 0 ; i < N ; ++i) {
  for (size_t j = i ; j < N ; ++j) {
   // matrix[i][j]
  }
}
```

In such cases, unrolling the outer loop can not only be beneficial by itself but can also enable the unrolling of the (copies of the) inner loop whose trip count will have been rendered runtime-independent, yielding significant performance benefits in some cases.

This commit adds a new static bonus to the unrolling threshold of outer loops for each of their inner loop whose trip count would go from runtime-dependent to runtime-independent if the outer loop were to be unrolled. This increases the likelihood to see interesting optimization opportunities in loop nests that have the particular structure outlined above.

Fixes SWDEV-485683.
</details>

lucas-rami committed:
Give cond. loop threshold bonus to outer loop in loop nests

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -148,8 +155,67 @@ void AMDGPUTTIImpl::getUnrollingPreferences(Loop *L, ScalarEvolution &SE,
-
   unsigned MaxBoost = std::max(ThresholdPrivate, ThresholdLocal);
+
+  if (llvm::PHINode *IV = L->getInductionVariable(SE)) {
```
Don't need llvm::

arsenm reviewed (commented):
I would expect this to be in the generic unroll logic, and not in this backend hook

arsenm commented on llvm/test/Transforms/LoopUnroll/AMDGPU/unroll-dependent-sub.ll:
```diff
@@ -0,0 +1,84 @@
+; RUN: opt -S -mtriple=amdgcn-- -passes=loop-unroll -debug-only=AMDGPUtti < %s 2>&1 | FileCheck %s
```
Better to check the actual output.

If you're going to check the debug output, might as well use -debug. But you also need REQUIRES: asserts

arsenm commented on llvm/test/Transforms/LoopUnroll/AMDGPU/unroll-dependent-sub.ll:
```diff
@@ -0,0 +1,84 @@
+end:                                                          ; preds = %outer.latch_exiting
+  ret void
+}
+
```
Can you add a case with a convergent operation that would be illegal to unroll, to make sure this doesn't accidentally override a correctness check

shiltian commented on llvm/test/Transforms/LoopUnroll/AMDGPU/unroll-dependent-sub.ll:
```diff
@@ -0,0 +1,84 @@
+; RUN: opt -S -mtriple=amdgcn-- -passes=loop-unroll -debug-only=AMDGPUtti < %s 2>&1 | FileCheck %s
+
+; For @dependent_sub_fullunroll, the threshold bonus should apply
+; CHECK: due to subloop's trip count becoming runtime-independent after unrolling
```
Also, please auto-gen the check lines.

lucas-rami committed:
Address reviewers' comments

shiltian commented:
I agree with @arsenm that this should go to a target independent part. If you don't want to break anything existing, the simplest solution would be to check the recurrence of the new argument, and only proceed when it is set.

lucas-rami committed:
Moved logic to target-independent analysis and improved it

lucas-rami commented:
I moved the optimization to the target-independent part of the loop unrolling pass. In the process, I significantly changed the way the analysis increases the likelihood of the loop being fully unrolled to be consistent with the way unroll cost estimation is generally done in the pass.

Instead of adding an absolute value to the unrolling threshold on encountering an optimizable subloop, my analysis now tries to estimate the per-iteration cost savings that fully unrolling a loop would yield, in an attempt to compare a more precise full unrolling cost against the threshold. The general idea is to find instructions in the loop which will be optimizable away once the loop IV is known. Subloops which may become unrollable if we were to unroll the outer loop yield additional cost savings. Simple control structure akin to if/then(/else) yield additional cost savings as well.

lucas-rami committed:
Fix clang-format

nikic commented:
This seems conceptually similar to something I tried a while ago (I have an old patch stub here: https://github.com/llvm/llvm-project/commit/26df05e27ce6c6742c6f523c926f4874ff5ca3ef). The problem I ran into, and why I didn't end up pursuing this further, was that if we more accurately model full unrolling cost, we end up unrolling more (duh) and from inspecting IR diffs the additional unrolls weren't always profitable. If we make cost modelling more accurate, we may have to reduce unrolling thresholds to compensate.

The other thing to watch out for, especially for your motivating example, is catastrophic unrolling. If you have this kind of nested loop and unroll the outer loop first, you can get quadratic unrolling. See https://github.com/llvm/llvm-project/issues/57865 for an example where simple input IR results in massive output IR that hangs the compiler and https://github.com/llvm/llvm-project/pull/67137 for a related patch. If I understand correctly, your patch is going to make this problem worse than it is right now, because now there will be an additional bonus to unrolling the outer loop.

lucas-rami commented:
Thanks a lot for adding context here, I see the danger of catastrophic unrolling and I now see that my patch will make it worse. My original motivation for this patch was in fact an AMDGPU kernel structurally similar to my motivating example where all loops need to be fully unrolled to achieve good performance (see discourse post [here](https://discourse.llvm.org/t/unexpected-peeling-decision-when-trying-to-eliminate-compares-inside-loop/82866/3) as well).

I think a solution would be, if there are subloops, to severely limit (or even nullify) the cost savings my analysis estimates if these subloops are too big or have big trip counts (what "too big" means to be defined somehow). If the latter cannot be determined, we would have to err on the "do not unroll" side. Do you think this would prevent catastrophic unrolling and still make the analysis a somewhat useful optimization?

lucas-rami committed:
Merge branch 'main' into unroll-thresold-bonus-to-nest

lucas-rami commented on llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp:
```diff
@@ -337,8 +336,239 @@ struct PragmaInfo {
+    Evaluated.insert(I);
+    if (isKnown(I))
+      return true;
+    if (!isa<BinaryOperator, CastInst, CmpInst>(I))
+      return false;
```
I may be too restrictive indeed. My rationale is that I only want to count instructions that would be folded by unrolling because all of their operands would be constants. I may just need to filter out instructions with side effects, phis, and branches?

lucas-rami committed:
Address feedback, almost NFC

The only functional change concerns the selection of potentially
foldable instructions inside loops ("foldable instructions" replaces
"known instructions" since the former seems like a more canonical term).
Instead of filtering-in instruction types that I think are foldable I
now filter-out instructions w/ side-effect, phis, and terminators.
SelectInst is special-cased because, unlike other instructions, it is
foldable as soon as its single "select operand" is known.

lucas-rami commented:
This is just addressing the inline remarks, I am still thinking of solutions on the larger issue of improving the heuristics to avoid catastrophic unrolling.
